text/0003-hash-indexed-object-store.md

```diff
@@ -0,0 +1,624 @@
+
+[future-possibilities]: #future-possibilities
+
+While not covered in detail in this RFC, these changes lay the groundwork for exciting future possiblities:
+
+### CDNs
+
+Since retrieving objects is just a matter of fetching by hash, we can leverage a CDN, for example via HTTP. Whenever a node needs an object `435abcf`, it can try to `GET http://casper.cdn.net/objects/435abcf` first. If found, we just saved the network the burden of sending the missing piece. This is not limited to the "official" CDN, validators are free to run one themselves, we may even add mechanisms to discover them.
+
+Open questions with this approach is who puts data on the CDN in the first place:
+
+* An easy first take is to upload daily snapshots of the whole chain ourselves. This already considerably reduces the burden on the network when new nodes are joining.
+* Another approach is to allow any node to push objects to the CDN. This might require a plausibility check (i.e. show that the object is reachable from genesis) and thus specialized software.
+* Clients could also upload WASM blocks to a CDN beforehand, making the actual deploy really small.
+
+Note that none of this is strictly required as all the CDN functionality is strictly optional with regards to the correct operation of the network.
+
+### Git-style pack files
+
+We may copy the idea of pack files, saving us some roundtrips. This is the first more complicated feature, but a node joining could easily happen to find a pack file that contains the
+blocks (without deploy contents) all the way up to genesis. It will then fetch other information as required.
+
+A pack is just a collection of objects, so importing one into a node should be (disk space not withstanding) strictly a benefit to the node. In the absence of CDNs, pack files can be used to speed up inter-node downloads.
+
+It is entirely possible to implement pack files as just a serialization of `HashMap<ObjectHash, Object>`.
+
+### Snapshots
+
+A feature currently on the roadmap becomes almost a non-issue with this approach; snapshotting or *checkpointing*, i.e. the ability to start a node without having to download all content from the network, requires no code modification in the sense that if objects are identified by immutable hashes, the local database can simply check if they are present to avoid redownloading them.
+
+Combined with packfiles, a snapshot or *checkpoint* of a node is just a specific block with all its dependencies in a single block file.
+
+### WASM deduplication
+
+A good thing to factor out from deploys is the WASM code blob, this may mean to do away with deploy headers entirely if the biggest part of the deploy is now given by reference instead of by value. With some changes to the deploy format or smart contract tooling, it may be possible to reference multiple WASM blobs in a single deploy; in a good design, an installed WASM blob would be found under the same hash in both the deploy and the global state.
```

Comment:
```suggestion
A good thing to factor out from Deploys is the WASM code blobs. We might also do away with DeployHeaders entirely if the biggest part of the Deploy is now given by reference instead of by value. 

With some changes to the deploy format or smart contract tooling, it may be possible to reference multiple WASM blobs in a single deploy; in a good design, an installed WASM blob would be found under the same hash in both the deploy and the global state.
```